### PR TITLE
GPU Package Initialization Fixes

### PR DESCRIPTION
**Pull request by wmbrownIntel**
**Summary**

Fixes potential issues and improves performance of initialization in the GPU package.

- Commit 1 fixes issue where shared main memory property is only set for nvidia devices.
- Commit 2 addresses potential issues with splitting accelerators into separate devices based on NUMA domain
- Commit 3 changes post_create() in lammps.cpp to prevent unnecessary reissue of "package" commands by a subsequent suffix command-line option to LAMMPS. There are rare cases where the default options to the gpu package might cause an issue.
- Commit 4 adds output of the selected platform to "screen" when multiple OpenCL platforms are available.
- Commit 5 makes parallel accelerator initialization the default when multiple MPI tasks share the same device. This greatly improves the initialization time for all APIs. In the past, this was serialized due to issues with both CUDA and OpenCL implementations using the same filename as a temporary cache, resulting in conflicts. This seems to be resolved in the major implementations for some time. The behavior can be reverted with -DLAL_SERIALIZE_INIT.
- Commit 6 documents some additional preprocessor flags in README and updates the oneAPI makefile (and adds some geryon preprocessor control that is currently only useful for development purposes). The intel compiler makefiles in MAKE/OPTIONS are also updated here with a flag to disable a deprecation remark for the classic compiler that we still use for better performance.

**Related Issue(s)**

**Author(s)**

Mike Brown, Intel

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

This modifies post_create() in lammps.cpp to process package commands first and prevent the suffix command from reissuing a package command.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: src/lammps.cpp [hunk 1]
```diff
@@ -858,10 +859,33 @@ void LAMMPS::post_create()
 {
   if (skiprunflag) input->one("timer timeout 0 every 1");
 
+  // Don't unnecessarily reissue a package command via suffix
```
Comment (akohlmey) [inline src/lammps.cpp:862]: The changes here break the `-pk pair/only on` flag. Unlike what is said in the comment, suffix *is* going to be NULL in that case (suffixp isn't) and thus this causes a segfault. I am looking into a workaround, but have not yet found one.

Comment (akohlmey) [reply src/lammps.cpp:862]: Workaround should be available now. This also fixes a related bug that has crept in unnoticed.

**Review by ndtrung81 — APPROVED** [review]
@wmbrownIntel thanks!

**Comment by akohlmey** [comment]
@wmbrownIntel FYI, I am still having issues with `-pk gpu 0 pair/only yes` not functioning correctly. My changes only fixed the case of one run statement, but with loops and using multiple runs, there are still issues. I suspect, the entire control mechanism for that feature needs to be redesigned. What was used so far, was a rather minimal approach.

In addition, there are problems with running multiple MPI tasks attached to the same GPU on my desktop with OpenCL and Intel HD 620 Graphics.
```$ mpirun -np 4 ./lmp -in ../bench/in.lj -sf gpu 
LAMMPS (15 Sep 2022)
  using 1 OpenMP thread(s) per MPI task
Could not create command queue on device: Intel(R) UHD Graphics [0x9b41]
application called MPI_Abort(MPI_COMM_WORLD, -1) - process 0
Could not create command queue on device: Intel(R) UHD Graphics [0x9b41]
application called MPI_Abort(MPI_COMM_WORLD, -1) - process 1
Could not create command queue on device: Intel(R) UHD Graphics [0x9b41]
application called MPI_Abort(MPI_COMM_WORLD, -1) - process 2
Could not create command queue on device: Intel(R) UHD Graphics [0x9b41]
application called MPI_Abort(MPI_COMM_WORLD, -1) - process 3
```
With other inputs (and using only 1 MPI rank) I also get errors like:
```
ERROR on proc 0: Neighbor list problem on the GPU. Try increasing the value of 'neigh_modify one' or the GPU neighbor list 'binsize'. (src/GPU/fix_gpu.cpp:337)
```
Which is worked around by setting the binsize value to what is chosen for the CPU explicitly for the GPU:
```
./lmp -in in.test -sf gpu -pk gpu 0 binsize 7.5
```

Any suggestions?

**Comment by wmbrownIntel** [comment]
@akohlmey Hopefully this [change](https://github.com/lammps/lammps/commit/ae993c83a0522a34f7fb3e2f19d322e39db6b5df) will fix the issue with multiple MPI tasks.

For the binsize issue, can I grab an input from somewhere (or generate from a benchmark) to reproduce?

For pair/only, I haven't been testing with this. The multiple run issue was a problem before this pull request? I can help; just trying to understand.

Thanks.

**Comment by akohlmey** [comment]
> @akohlmey Hopefully this [change](https://github.com/lammps/lammps/commit/ae993c83a0522a34f7fb3e2f19d322e39db6b5df) will fix the issue with multiple MPI tasks.

Unfortunately, no. It doesn't stop right away, but now I am getting a segmentation fault (e.g. with bench/in.lj) when I use more than 1 MPI rank. The segfault is deep inside the OpenCL code:
```
(gdb) where
#0  ___pthread_once (once_control=0x2d72, init_routine=0x7f3c9a2da770 <std::__once_proxy()>) at pthread_once.c:139
#1  0x00007f3c94246840 in __gthread_once (__func=<optimized out>, __once=0x2d72) at /usr/include/c++/12/x86_64-redhat-linux/bits/gthr-default.h:700
#2  std::call_once<NEO::Drm::isVmBindAvailable()::<lambda()> > (__f=..., __once=...) at /usr/include/c++/12/mutex:859
#3  NEO::Drm::isVmBindAvailable (this=0x2bf2) at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/shared/source/os_interface/linux/drm_neo.cpp:1267
#4  0x00007f3c94235770 in NEO::BufferObject::wait (this=this@entry=0x2bf2890, timeoutNs=timeoutNs@entry=-1)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/shared/source/os_interface/linux/drm_buffer_object.cpp:86
#5  0x00007f3c9423b8bb in NEO::DrmMemoryManager::cleanOsHandles (this=0x1db7e20, handleStorage=..., rootDeviceIndex=<optimized out>)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/shared/source/os_interface/linux/drm_memory_manager.cpp:1038
#6  0x00007f3c9423be8b in NEO::DrmMemoryManager::freeGraphicsMemoryImpl (this=0x1db7e20, gfxAllocation=0x1fc5650, isImported=<optimized out>)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/shared/source/os_interface/linux/drm_memory_manager.cpp:924
#7  0x00007f3c941c7e70 in NEO::InternalAllocationStorage::freeAllocationsList (this=0x1ba1e60, waitTaskCount=34, allocationsList=...)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/shared/source/memory_manager/internal_allocation_storage.cpp:58
#8  0x00007f3c941c809a in NEO::InternalAllocationStorage::cleanAllocationList (this=0x1ba1e60, waitTaskCount=34, allocationUsage=0)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/shared/source/memory_manager/internal_allocation_storage.cpp:41
#9  0x00007f3c9415567c in NEO::CommandStreamReceiver::waitForTaskCountAndCleanAllocationList (this=0x1dbd790, requiredTaskCount=<optimized out>, 
    allocationUsage=allocationUsage@entry=0) at /usr/include/c++/12/bits/unique_ptr.h:191
#10 0x00007f3c941556ab in NEO::CommandStreamReceiver::waitForTaskCountAndCleanTemporaryAllocationList (this=<optimized out>, requiredTaskCount=<optimized out>)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/shared/source/command_stream/command_stream_receiver.cpp:200
#11 0x00007f3c93cc9bd1 in NEO::CommandQueue::waitUntilComplete (this=this@entry=0x1bd6ae0, gpgpuTaskCountToWait=<optimized out>, copyEnginesToWait=..., 
    flushStampToWait=146028888080, useQuickKmdSleep=useQuickKmdSleep@entry=false, cleanTemporaryAllocationList=true, skipWait=false)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/opencl/source/command_queue/command_queue.cpp:382
#12 0x00007f3c93cca6f5 in NEO::CommandQueue::waitForAllEngines (this=this@entry=0x1bd6ae0, blockedQueue=blockedQueue@entry=true, printfHandler=printfHandler@entry=0x0, 
    cleanTemporaryAllocationsList=cleanTemporaryAllocationsList@entry=true)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/opencl/source/command_queue/command_queue.cpp:1163
#13 0x00007f3c93d836a1 in NEO::CommandQueue::waitForAllEngines (printfHandler=0x0, blockedQueue=true, this=0x1bd6ae0)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/opencl/source/command_queue/command_queue.h:217
#14 NEO::CommandQueueHw<NEO::Gen9Family>::finish (this=0x1bd6ae0) at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/opencl/source/command_queue/finish.h:24
#15 0x00007f3c93ccbc9e in NEO::CommandQueue::cpuDataTransferHandler (this=0x1bd6ae0, transferProperties=..., eventsRequest=..., retVal=<optimized out>)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/opencl/source/command_queue/cpu_data_transfer_handler.cpp:97
#16 0x00007f3c93cc7913 in NEO::CommandQueue::enqueueMapBuffer (this=this@entry=0x1bd6ae0, buffer=buffer@entry=0x1f98f10, blockingMap=<optimized out>, mapFlags=2, 
    offset=<optimized out>, size=<optimized out>, numEventsInWaitList=0, eventWaitList=0x0, event=0x0, errcodeRet=@0x7ffd66d661c4: 0)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/opencl/source/command_queue/command_queue.cpp:659
#17 0x00007f3c93c9f84a in clEnqueueMapBuffer (commandQueue=<optimized out>, buffer=<optimized out>, blockingMap=<optimized out>, mapFlags=<optimized out>, offset=<optimized out>, 
    cb=<optimized out>, numEventsInWaitList=<optimized out>, eventWaitList=<optimized out>, event=<optimized out>, errcodeRet=<optimized out>)
    at /usr/src/debug/intel-compute-runtime-22.34.24023-1.fc36.x86_64/opencl/source/api/api.cpp:3248
#18 0x00007f3c9d03acd1 in ucl_opencl::_host_alloc<ucl_opencl::UCL_H_Vec<float> > (kind2=<optimized out>, kind=UCL_WRITE_ONLY, n=524848, dev=..., mat=...)
    at /home/akohlmey/compile/lammps/lib/gpu/geryon/ocl_memory.h:168
#19 ucl_opencl::UCL_H_Vec<float>::alloc (kind2=UCL_READ_ONLY, kind=UCL_WRITE_ONLY, device=..., cols=131212, this=0x7f3c9d8fc3c8 <LAMMPS_AL::global_device+8>)
    at /home/akohlmey/compile/lammps/lib/gpu/geryon/ucl_h_vec.h:103
#20 ucl_opencl::_ucl_s_obj_help<1>::alloc<ucl_opencl::UCL_H_Vec<float>, ucl_opencl::UCL_D_Vec<float>, ucl_opencl::UCL_H_Vec<float> > (host=..., device=..., cols=131212, acc=..., 
    kind1=kind1@entry=UCL_WRITE_ONLY, kind2=kind2@entry=UCL_READ_ONLY, _buffer=...) at /home/akohlmey/compile/lammps/lib/gpu/geryon/ucl_s_obj_help.h:39
#21 0x00007f3c9d03d35e in ucl_opencl::UCL_Vector<float, float>::alloc (kind2=UCL_READ_ONLY, kind1=UCL_WRITE_ONLY, acc=..., cols=<optimized out>, this=0x7f3c9d8fc3c8 <LAMMPS_AL::global_device+8>) at /home/akohlmey/compile/lammps/lib/gpu/geryon/ucl_vector.h:94
#22 LAMMPS_AL::Atom<float, double>::alloc (this=this@entry=0x7f3c9d8fc3c8 <LAMMPS_AL::global_device+8>, nall=nall@entry=29821) at /home/akohlmey/compile/lammps/lib/gpu/lal_atom.cpp:100
#23 0x00007f3c9d043573 in LAMMPS_AL::Atom<float, double>::resize (success=@0x7ffd66d6661b: true, nall=29821, this=0x7f3c9d8fc3c8 <LAMMPS_AL::global_device+8>) at /home/akohlmey/compile/lammps/lib/gpu/lal_atom.h:87
#24 LAMMPS_AL::BaseAtomic<float, double>::resize_atom (success=@0x7ffd66d6661b: true, nall=29821, inum=16000, this=0x7f3c9d902a00 <LJLMF>) at /home/akohlmey/compile/lammps/lib/gpu/lal_base_atomic.h:65
#25 LAMMPS_AL::BaseAtomic<float, double>::build_nbor_list (this=this@entry=0x7f3c9d902a00 <LJLMF>, inum=16000, host_inum=0, nall=nall@entry=29821, host_x=host_x@entry=0x297d7a0, host_type=host_type@entry=0x2de1aa0, sublo=0x7ffd66d66630, subhi=0x7ffd66d66650, tag=0x2c6a8c0, nspecial=0x0, special=0x0, success=@0x7ffd66d6661b: true) at /home/akohlmey/compile/lammps/lib/gpu/lal_base_atomic.cpp:168
#26 0x00007f3c9d043888 in LAMMPS_AL::BaseAtomic<float, double>::compute (this=this@entry=0x7f3c9d902a00 <LJLMF>, ago=0, inum_full=inum_full@entry=16000, nall=29821, host_x=0x297d7a0, host_type=0x2de1aa0, sublo=0x7ffd66d66630, subhi=0x7ffd66d66650, tag=0x2c6a8c0, nspecial=0x0, special=0x0, eflag_in=true, vflag_in=true, eatom=false, vatom=false, host_start=@0x7ffd66d6661c: 16000, ilist=0x7ffd66d66620, jnum=0x7ffd66d66628, cpu_time=0, success=@0x7ffd66d6661b: true) at /home/akohlmey/compile/lammps/lib/gpu/lal_base_atomic.cpp:280
#27 0x00007f3c9d013bc8 in ljl_gpu_compute_n (ago=<optimized out>, inum_full=inum_full@entry=16000, nall=<optimized out>, host_x=<optimized out>, host_type=<optimized out>, sublo=sublo@entry=0x7ffd66d66630, subhi=0x7ffd66d66650, tag=0x2c6a8c0, nspecial=0x0, special=0x0, eflag=true, vflag=true, eatom=false, vatom=false, host_start=@0x7ffd66d6661c: 16000, ilist=0x7ffd66d66620, jnum=0x7ffd66d66628, cpu_time=<optimized out>, success=@0x7ffd66d6661b: true) at /home/akohlmey/compile/lammps/lib/gpu/lal_lj_ext.cpp:123
#28 0x00007f3c9cdde936 in LAMMPS_NS::PairLJCutGPU::compute (this=0x2192de0, eflag=1, vflag=1) at /home/akohlmey/compile/lammps/src/GPU/pair_lj_cut_gpu.cpp:99
#29 0x00007f3c9bfca992 in LAMMPS_NS::Verlet::setup (this=0x1ba8ae0, flag=1) at /home/akohlmey/compile/lammps/src/verlet.cpp:140
#30 0x00007f3c9bf77b97 in LAMMPS_NS::Run::command (this=0x296c700, narg=<optimized out>, arg=<optimized out>) at /home/akohlmey/compile/lammps/src/run.cpp:171
#31 0x00007f3c9be17f64 in LAMMPS_NS::Input::execute_command (this=0x1b98ba0) at /home/akohlmey/compile/lammps/src/input.cpp:847
#32 0x00007f3c9be1833e in LAMMPS_NS::Input::file (this=0x1b98ba0) at /home/akohlmey/compile/lammps/src/input.cpp:302
#33 0x00000000004012ae in main (argc=<optimized out>, argv=<optimized out>) at /home/akohlmey/compile/lammps/src/main.cpp:76
```

> For the binsize issue, can I grab an input from somewhere (or generate from a benchmark) to reproduce?

I was using the files from this discussion: https://matsci.org/t/lammps-loop-fails-in-gpu-accelerate/44922
It seems this issue already existed before this pull request.

> For pair/only, I haven't been testing with this. The multiple run issue was a problem before this pull request? I can help; just trying to understand.

I think the pair/only issue is a more general problem. It was broken before the GPU init PR. Then I "fixed" it for a single run case, but I suspect that there is a fundamental design issue in how this was implemented and I have to re-think it. So please don't invest any effort in this. I will be looking into it tomorrow and have already some ideas to try out.

> Thanks.

**Comment by wmbrownIntel** [comment]
@akohlmey Third time is the [charm](https://github.com/lammps/lammps/commit/54840211869de5bb4b2ae9322e6f0813ecb3a086)? 

The timing disable at multiple ranks left a spot in the initial GPU overhead estimation where the events/markers used for timing could be destroyed before the accelerator was done. Hopefully this was causing corruption leading to your crash in an unrelated spot. My test machines have lower overheads, so the issue escaped.

It will probably take a little time for me to get to the binsize issue, but hopefully next week.

**Comment by akohlmey** [comment]
> @akohlmey Third time is the [charm](https://github.com/lammps/lammps/commit/54840211869de5bb4b2ae9322e6f0813ecb3a086)?

Almost. I can now run several test/bench inputs with `mpirun -np 4` without error or segfault. However, the test input with the binsize problem runs for a while and then (at random times?) stops with this kind of error:
```
Per MPI rank memory allocation (min/avg/max) = 16.65 | 16.67 | 16.72 Mbytes
   Step         S/CPU         Atoms         Temp          PotEng         KinEng         TotEng        c_xtemp         v_Nch4        v_Nch4f        v_Nch4ad   
         5   0                  11147   333           -2072412.9      575.05191     -2071837.8      333            1739           1061           678          
OpenCL error in file '/home/akohlmey/compile/lammps/lib/gpu/geryon/ocl_timer.h' in line 118 : -14.
application called MPI_Abort(MPI_COMM_WORLD, -1) - process 1
```
 
> The timing disable at multiple ranks left a spot in the initial GPU overhead estimation where the events/markers used for timing could be destroyed before the accelerator was done. Hopefully this was causing corruption leading to your crash in an unrelated spot. My test machines have lower overheads, so the issue escaped.
> 
> It will probably take a little time for me to get to the binsize issue, but hopefully next week.

OK. In the meantime, I have cherry-picked both commits into PR #3485 since they are definite improvements. Thanks for that.

**Comment by akohlmey** [comment]
@wmbrownIntel I think I have resolved the `-pk gpu 0 pair/local yes` issue in PR #3485.
It was never set up to preserve the required settings and pointers across a `clear` command. This is now supported.

**Pull request merged**